model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java

```diff
@@ -374,24 +374,30 @@ public void validateAndUpdateSchema(KeycloakSession session, Class<?> modelType)
 
         if (this.validatedModelNames.add(modelName)) {
         */
-        if (this.validatedModels.add(modelType)) {
-            Connection connection = getConnection();
-            try {
-                if (logger.isDebugEnabled()) printOperationalInfo(connection);
-
-                MapJpaUpdaterProvider updater = session.getProvider(MapJpaUpdaterProvider.class);
-                MapJpaUpdaterProvider.Status status = updater.validate(modelType, connection, config.get("schema"));
-
-                if (!status.equals(VALID)) {
-                    update(modelType, connection, session);
-                }
-            } finally {
-                if (connection != null) {
+        if (!this.validatedModels.contains(modelType)) {
+            if (!this.validatedModels.contains(modelType)) {
+                //noinspection SynchronizationOnLocalVariableOrMethodParameter
+                synchronized (modelType) {
```

Comment:
Sync should be on an object owned by this `JpaMapStorageProviderFactory.java` instance. How about declaring a `private final ConcurrentHashMap<Class<?>, Object> SYNC_MODELS = new ConcurrentHashMap()` and using it here?

```suggestion
                synchronized (SYNC_MODELS.computeIfAbsent(modelClass, mc -> new Object())) {
```

Reply:
My IDE had a similar suggestion, therefore the "noinspection SynchronizationOnLocalVariableOrMethodParameter" comment. 

This could be debatable, as `modelType` is a class, and classes are singletons inside a classloader context. That's why I kept it as is.

Reply:
I'd prefer to have it done so that the lock is done locally within the instance, if only to allow for concurrent initialization for say `UserModel` schema from several realms concurrently.